libraries/stdlib/osl/mx_funcs.h

```diff
@@ -469,6 +469,303 @@ vector4 fBm( point position, int octaves, float lacunarity, float diminish, stri
+/// TODO:
+/// The right shift operation below should be an unsigned int right shift
+/// in order to match the hash used in GLSL and MDL. This is not supported
+/// in OSL so for now the hashes are not the same and the results produced
+/// are not identical between GLSL/MDL/OSL.
+///
+int mx_rotl32(int x, int k)
+{
+    return (x<<k) | (x>>(32-k));
+}
+
+// Mix up and combine the bits of a, b, and c (doesn't change them, but
+// returns a hash of those three original values).
+int mx_bjfinal(int _a, int _b, int _c)
+{
+    int a = _a;
+    int b = _b;
+    int c = _c;
+    c ^= b; c -= mx_rotl32(b,14);
+    a ^= c; a -= mx_rotl32(c,11);
+    b ^= a; b -= mx_rotl32(a,25);
+    c ^= b; c -= mx_rotl32(b,16);
+    a ^= c; a -= mx_rotl32(c,4);
+    b ^= a; b -= mx_rotl32(a,14);
+    c ^= b; c -= mx_rotl32(b,24);
+    return c;
+}
+
+int mx_hash_int(int x, int y)
+{
+    int len = 2;
+    int a, b, c;
+    a = b = c = int(0xdeadbeef) + (len << 2) + 13;
+    a += x;
+    b += y;
+    return mx_bjfinal(a, b, c);
+}
+
+int mx_hash_int(int x, int y, int z)
+{
+    int len = 3;
+    int a, b, c;
+    a = b = c = int(0xdeadbeef) + (len << 2) + 13;
+    a += x;
+    b += y;
+    c += z;
+    return mx_bjfinal(a, b, c);
+}
+
+struct int3
+{
+    int x;
+    int y;
+    int z;
+};
+
+int3 mx_hash_int3(int x, int y)
+{
+    int h = mx_hash_int(x, y);
+    // we only need the low-order bits to be random, so split out
+    // the 32 bit result into 3 parts for each channel
+    int3 result;
+    result.x = (h      ) & 0xFF;
+    result.y = (h >> 8 ) & 0xFF;
+    result.z = (h >> 16) & 0xFF;
+    return result;
+}
+
+int3 mx_hash_int3(int x, int y, int z)
+{
+    int h = mx_hash_int(x, y, z);
+    // we only need the low-order bits to be random, so split out
+    // the 32 bit result into 3 parts for each channel
+    int3 result;
+    result.x = (h      ) & 0xFF;
+    result.y = (h >> 8 ) & 0xFF;
+    result.z = (h >> 16) & 0xFF;
+    return result;
+}
+
+float mx_8bits_to_01(int bits) { return float(bits) / float(0xff); }
+
+float mx_worley_distance2(vector2 p, int x, int y, int xoff, int yoff, float jitter, int metric)
+{
+    int3 hash = mx_hash_int3(x+xoff, y+yoff);
+    vector2 off = vector2(mx_8bits_to_01(hash.x), mx_8bits_to_01(hash.y));
+
+    off -= 0.5;
+    off *= jitter;
+    off += 0.5;
+
+    vector2 cellpos = vector2(float(x), float(y)) + off;
+    vector2 diff = cellpos - p;
+    if (metric == 2)
+        return abs(diff.x) + abs(diff.y);       // Manhattan distance
+    if (metric == 3)
+        return max(abs(diff.x), abs(diff.y));   // Chebyshev distance
+    // Either Euclidian or Distance^2
+    return dot(diff, diff);
+}
+
+float mx_worley_distance3(vector p, int x, int y, int z, int xoff, int yoff, int zoff, float jitter, int metric)
+{
+    int3 hash = mx_hash_int3(x+xoff, y+yoff, z+zoff);
+    vector off = vector(mx_8bits_to_01(hash.x),
+                        mx_8bits_to_01(hash.y),
+                        mx_8bits_to_01(hash.z));
```

Comment:
Note that the very definition of cellnoise (which is in MaterialX, and I presume is identical to that of OSL) serves the same purpose as this -- returning a deterministic uncorrelated [0,1] range value for each "cell" [i,i+1)^N in the volume. And that appears to be exactly what is at the heart of this construct and how it's used.

I feel like maybe an opportunity was missed here, where the specification of the worley functions in MX could have simply stipulated that the random point positions within each integer lattice cell were *by definition* identical to the cellnoise results for the contents of the cell. That would allow the top level of the mx_worley_noise_TYPE functions to call cellnoise to get the offsets and thus eliminate the majority of the lines of code in this module, as well as guaranteeing use of any carefully optimized cellnoise implementations in the {CPU, SIMD, or GPU} inside OSL or any other languages. Expanding the entire implementation out in shader source code down to every last bit twiddle seems very unlikely to be as efficient as the versions we've already optimized behind the cellnoise implementations.

The whole idea of cellnoise, back from when I first implemented it in PRMan (circa 1996?) was to be a "primitive function" from which you could build up higher level things like worley patterns, without having to get lost in the bit-by-bit details. This is literally what it was born to do!

Is it too late to allow for changes in the canonical pattern in order to pave the way for all the benefits of having cellnoise and worley overlap in their implementation and results?

Reply:
That's a great observation, and it's still possible to align the definitions of Worley noise and cellnoise in the way that you propose.  I'd be interested in thoughts from @code-monkey and @niklasharrysson on this idea, and I'll plan to bring this up at today's TSC meeting.

Reply:
Great points @lgritz! And I also agree it would be nice to align this. I think we may need some input from @code-monkey about the implementation details, and if/how the use of cellnoise would differ from the GLSL implementation.

If it can't be done that way we have the other option of exposing the 2D and 3D hash functions as you mentioned above.

Reply:
In MaterialX, there are only versions of cellnoise which return a single float value.  They boil down to a simple hash function on the cell location.  It would be good to extend the GLSL noise library to include versions that returned a vec2/vec3/vec4 for the "offset" inside the cell.  If OSL versions already exist for these functions, it would be ideal to use them for multiple reasons.  These would be perfect for the Worley noise implementation (as Larry points out).
Maybe a future spec of MaterialX could also expose the cellnoise functions that return vector types.

The MaterialX  GLSL implementation for Worley Noise would need to be patched with the OSL hashing functions.  It's not too late to change the reference implementation (IMO).